android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -440,6 +440,30 @@ class GeckoEngineSession(
         }
     }
 
+    /**
+     * See [EngineSession.hasCookieBannerRuleForSession]
+     */
+    override fun hasCookieBannerRuleForSession(
+        onSuccess: (Boolean) -> Unit,
+        onError: (Throwable) -> Unit,
+    ) {
+        geckoSession.hasCookieBannerRuleForBrowsingContextTree().then(
+            { response ->
+                if (response == null) {
+                    logger.error("No result from GeckoView for cookie banner rule.")
+                    return@then GeckoResult<Boolean>()
```

Comment:
Should we call `onError()` for this case? As it looks like if hit this section neither `onError()` or `onSuccess()` will be called?

Reply:
Could we cover this flow with test? :)

Reply:
> Should we call `onError()` for this case? As it looks like if hit this section neither `onError()` nor `onSuccess()` will be called?

I don't think so. First of, when we implemented `checkForFormData()` [here](https://github.com/mozilla-mobile/firefox-android/pull/264/files) we didn't. Second, `onError` is for handling the error that comes from the API. In this case we are handling a legit response (which is not that likely, I added this null check mostly to get rid of the `Boolean?` to improve ergonomics of the API)

Reply:
> Could we cover this flow with test? :)

What exactly do you hope to test there? The behaviour of the logger? In the patch I mentioned above similar code path is also not tested as far as I can tell.

Reply:
> What exactly do you hope to test there? The behaviour of the logger? In the patch I mentioned above similar code path is also not tested as far as I can tell.

I was looking to exercise how we are notifying the caller when this case happens see https://github.com/mozilla-mobile/firefox-android/pull/1060#discussion_r1127028167.

I'm bit worry to confirm if we are in a state, *where we don't notify the caller, this function will block the door hanger from showing until we get the value* see https://github.com/mozilla-mobile/firefox-android/blob/22563c8ae26b6de2b0eabd9b1c5f7ddbd1012038/fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt#L379

Reply:
How is the caller notified in the case of `checkForFormData()`? I think the way in both cases the caller gets notified is `onSuccess` is not getting called (+ in the case of `checkForFormData()` we don't notify observers either). Why would the function block? It returns in the case of `null`: `return@then GeckoResult<Boolean>()`, but maybe you have a better understanding?

Reply:
I mean, I certainly can add the test case to assert we don't call `onSuccess` in the case of `null` if you think that's valuable

Reply:
For the `checkForFormData()`I think we missed that and we should file a bug (**I'll do it in a bit**), that functionality is not a user facing one, it's only use to [report telemetry](https://github.com/mozilla-mobile/fenix/pull/28697/files), it will fail silently.  On the other hand, `hasCookieBannerRuleForSession` will be use directly on the UI if we don't report all cases, we may miss some part of the UI.

`hasCookieBannerRuleForSession` will be use to construct this UI.
![image](https://user-images.githubusercontent.com/773158/223237451-317d7353-0b87-48cc-8cd9-d4c5b47a356d.png)

Reply:
Bug for [checkForFormData()](https://bugzilla.mozilla.org/show_bug.cgi?id=1820803)